pkg/chartsync/git.go

```diff
@@ -0,0 +1,337 @@
+	if err != nil {
+		return GitChartDir{}, err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), c.config.GitTimeout)
+	defer cancel()
+	export, err := repo.Export(ctx, s.head)
+	if err != nil {
+		return GitChartDir{}, ChartUnavailableError{err}
+	}
+
+	return GitChartDir{export, s.head}, nil
+}
+
+// Delete cleans up the source reference for the given `v1.HelmRelease`,
+// this includes the mirror if there is no reference to it from sources.
+// It returns a boolean indicating a successful removal (`true` if so,
+// `false` otherwise).
+func (c *GitChartSync) Delete(hr *v1.HelmRelease) bool {
+	c.sourcesMu.Lock()
+	defer c.sourcesMu.Unlock()
+
+	// Attempt to get the source from store.
+	source, ok := c.sources[hr.ResourceID().String()]
+	if ok {
+		// Remove the in store source.
+		delete(c.sources, hr.ResourceID().String())
+
+		if hrs, err := c.helmReleasesForMirror(source.mirror); err == nil && len(hrs) == 0 {
+			// The mirror is no longer in use by any source;
+			// stop and delete the mirror.
+			c.mirrors.StopOne(source.mirror)
+		}
+	}
+	return ok
+}
+
+// SyncMirrors instructs all git mirrors to sync from their respective
+// upstreams.
+func (c *GitChartSync) SyncMirrors() {
+	c.logger.Log("info", "starting sync of git mirrors")
+	for _, err := range c.mirrors.RefreshAll(c.config.GitTimeout) {
+		c.logger.Log("error", ErrMirrorSync.Error(), "err", err)
+	}
+	c.logger.Log("info", "finished syncing git mirror")
+}
+
+// processChangedMirror syncs all given `v1.HelmRelease`s with the
+// mirror we received a change signal for and schedules a release,
+// but only if the sync indicated the change was relevant.
+func (c *GitChartSync) processChangedMirror(mirror string, repo *git.Repo, hrs []*v1.HelmRelease) {
+	for _, hr := range hrs {
+		if _, ok, _ := c.sync(hr, mirror, repo); ok {
+			cacheKey, err := cache.MetaNamespaceKeyFunc(hr.GetObjectMeta())
+			if err != nil {
+				continue // this should never happen
+			}
+			// Schedule release sync by adding it to the queue.
+			c.releaseQueue.AddRateLimited(cacheKey)
+		}
+	}
+}
+
+func (c *GitChartSync) get(hr *v1.HelmRelease) (sourceRef, bool) {
+	c.sourcesMu.RLock()
+	defer c.sourcesMu.RUnlock()
+	if s, ok := c.sources[hr.ResourceID().String()]; ok && s.forHelmRelease(hr) {
+		return s, ok
+	}
+	return sourceRef{}, false
+}
+
+func (c *GitChartSync) store(hr *v1.HelmRelease, s sourceRef) {
+	c.sourcesMu.Lock()
+	c.sources[hr.ResourceID().String()] = s
+	c.sourcesMu.Unlock()
+}
+
+// sync synchronizes the record we have for the given `v1.HelmRelease`
+// with the given mirror. It always updates the HEAD record in the
+// `sourceRef`, but only returns `true` if the update was relevant for
+// the release (e.g. a change in git the chart source path, or a new
+// record). In case of failure it returns an error.
+func (c *GitChartSync) sync(hr *v1.HelmRelease, mirrorName string, repo *git.Repo) (sourceRef, bool, error) {
+	source := hr.Spec.GitChartSource
+	if source == nil {
```

Comment:
It's strange that the method needs to check for the `GitChartSource` and also provide a `mirrorName`, which should be a subproduct of the former).

Reply:
I would change the interface to `sync(hrID resource.ID, mirrorName string, repo *git.Repo) ` 

But, as it is, that's no possible, because `sync` also consistency-checks that the mirror we have for the Helm release (in `get()`s invocation to `forHelmRelease()`).

However, I don't think that consistency-checking is the direct responsibility of `sync()` (or at least it makes it harder to understand what's going on since it's buried when invoking `get()`).

I think that we should take car of divergences explicitly, while also garbage collecting orphan mirrors (i.e. https://github.com/fluxcd/helm-operator/pull/99/files#r347533178 )

Reply:
There is another reason this is not possible, without the `v1.HelmRelease` or a copy of the `v1.GitChartSource`, it is not possible to detect changes as done on L286, nor is it possible to construct a new `sourceRef` as done on L265.